### PR TITLE
Implement pausing and resuming sessions

### PR DESCRIPTION
**Pull request by clementetb**
This PR allows to control a SyncSession lifecycle by implementing the methods `pause` and `resume`.

They are required to be able to simulate a client reset.

### CODE REVIEW

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt [hunk 1]
```diff
@@ -68,6 +68,34 @@ public interface SyncSession {
      */
     public suspend fun uploadAllLocalChanges(timeout: Duration = Duration.INFINITE): Boolean
 
+    /**
+     * The current SyncSession state. See [SyncSessionState] for more details about each state.
```
Comment (cmelchior) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt:72]:
```suggestion
     * The current session state. See [SyncSessionState] for more details about each state.
```

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt [hunk 2]
```diff
@@ -68,6 +68,34 @@ public interface SyncSession {
...
+     */
+    public val state: SyncSessionState
+
+    /**
+     * Pauses any synchronization with the Realm Object Server until the Realm is re-opened again
+     * after fully closing it.
+     * <p>
+     * Synchronization can be re-activated by calling [resume] again.
+     * <p>
+     * If the session is already [SyncSessionState.INACTIVE], calling this method will do nothing.
```
Comment (cmelchior) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt:82]:
```suggestion
     * Pauses synchronization with Atlas until the Realm is closed and re-opened again.
     * 
     * Synchronization can also be re-activated by calling [resume].
     *
     * If the session is already [SyncSessionState.INACTIVE], calling this method will do nothing.
```

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt [hunk 3]
```diff
@@ -68,6 +68,34 @@ public interface SyncSession {
...
+    /**
+     * Attempts to resume the session and activate synchronization with the Realm Object Server.
+     * <p>
+     * This happens automatically when opening the Realm instance, so doing it manually should only
+     * be needed if the session was paused using [pause].
+     * <p>
+     * If the session was already [SyncSessionState.ACTIVE], calling this method will do nothing.
+     * <p>
+     * If the session state is [SyncSessionState.DYING], the session will be moved back to
+     * [SyncSessionState.ACTIVE].
```
Comment (cmelchior) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt:95]:
```suggestion
     * Attempts to resume the session and activate synchronization with Atlas.
     *
     * This happens automatically when opening the Realm, so doing it manually should only
     * be needed if the session was paused using [pause].
     * 
     * If the session was already [SyncSessionState.ACTIVE], calling this method will do nothing.
     * 
     * If the session state is [SyncSessionState.DYING], the session will be moved back to
     * [SyncSessionState.ACTIVE].
```

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt [hunk 4]
```diff
@@ -86,3 +114,13 @@ public interface SyncSession {
         public fun onError(session: SyncSession, error: SyncException)
     }
 }
+
+/**
+ * The possible states for [SyncSession] to be.
+ */
+public enum class SyncSessionState {
```
Comment (cmelchior) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt:121]: There is a lot of documentation in Realm Java we can move over for these.

Comment (cmelchior) [reply packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncSession.kt:121]: I would probably move into`SyncSession.State`, similar to `User.State` and how it is done in Realm Java.

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncSessionTests.kt [hunk 5]
```diff
@@ -87,6 +88,25 @@ class SyncSessionTests {
         }
     }
 
+    @Test
+    fun sessionPauseAndResume() {
```
Comment (cmelchior) [inline test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncSessionTests.kt:92]: There are a few more tests in https://github.com/realm/realm-java/blob/f0eca830527cd8255ce8fb9bcc83a545a940f6d3/realm/realm-library/src/syncIntegrationTest/kotlin/io/realm/SyncSessionTests.kt#L729 we should probably move as well (i.e. calling pause/resume multiple times)

**Review by cmelchior — CHANGES_REQUESTED** [review]
Nice to get this in. I added some comments for some smaller adjustments I think we should make and tests from Java we should port...and also the CHANGELOG ;)

**Review by cmelchior — APPROVED** [review]
Looks good to me. If CI is happy 👍

Still need a changelog update though ;)

**Comment by clementetb** [comment]
🙈

**Pull request merged**